dbos/_dbos_config.py

```diff
@@ -323,75 +314,70 @@ def process_config(
-        if rollback:
-            dbconfig["rollback"] = cast(List[str], rollback)
-        data["database"] = dbconfig
-
-    if "app_db_name" not in data["database"] or not (data["database"]["app_db_name"]):
-        data["database"]["app_db_name"] = _app_name_to_db_name(data["name"])
-
-    connection_passed_in = data["database"].get("hostname", None) is not None
-
-    dbos_dbport: Optional[int] = None
-    dbport_env = os.getenv("DBOS_DBPORT")
-    if dbport_env:
-        try:
-            dbos_dbport = int(dbport_env)
-        except ValueError:
-            pass
-
-    data["database"]["hostname"] = (
-        os.getenv("DBOS_DBHOST") or data["database"].get("hostname") or "localhost"
-    )
-
-    data["database"]["port"] = dbos_dbport or data["database"].get("port") or 5432
-    data["database"]["username"] = (
-        os.getenv("DBOS_DBUSER") or data["database"].get("username") or "postgres"
-    )
-    data["database"]["password"] = (
-        os.getenv("DBOS_DBPASSWORD")
-        or data["database"].get("password")
-        or os.environ.get("PGPASSWORD")
-        or "dbos"
-    )
-
-    if not data["database"].get("app_db_pool_size"):
-        data["database"]["app_db_pool_size"] = 20
-    if not data["database"].get("sys_db_pool_size"):
-        data["database"]["sys_db_pool_size"] = 20
-    if not data["database"].get("connectionTimeoutMillis"):
-        data["database"]["connectionTimeoutMillis"] = 10000
-
+    # Handle admin server config
     if not data.get("runtimeConfig"):
         data["runtimeConfig"] = {
             "run_admin_server": True,
         }
     elif "run_admin_server" not in data["runtimeConfig"]:
         data["runtimeConfig"]["run_admin_server"] = True
 
-    # Pretty-print where we've loaded database connection information from, respecting the log level
-    if not silent and logs["logLevel"] == "INFO" or logs["logLevel"] == "DEBUG":
-        d = data["database"]
-        conn_string = f"postgresql://{d['username']}:*****@{d['hostname']}:{d['port']}/{d['app_db_name']}"
-        if os.getenv("DBOS_DBHOST"):
-            print(
-                f"[bold blue]Loading database connection string from debug environment variables: {conn_string}[/bold blue]"
+    isDebugMode = os.getenv("DBOS_DBHOST") is not None
+
+    # Ensure database dict exists
+    data.setdefault("database", {})
+    configure_db_engine_parameters(data["database"])
+
+    # Database URL resolution
+    if data.get("database_url") is not None:
+        # Parse the db string and check required fields
+        assert data["database_url"] is not None
+        url = make_url(data["database_url"])
+        required_fields = [
+            ("username", "Username must be specified in the connection URL"),
+            ("password", "Password must be specified in the connection URL"),
+            ("host", "Host must be specified in the connection URL"),
+            ("database", "Database name must be specified in the connection URL"),
```

Comment:
I am hesitating to make the port mandatory too. I'm not sure what happens if you get a managed postgres and are provided with a URL with no port. Unclear that said unprovided port is 5432.

Reply:
Also in a future PR I'll make that factor out this validation and use it in other places, like the client's constructor.